src/lib/consensus/proof_of_stake.ml

```diff
@@ -3916,6 +3914,711 @@ let%test_module "Proof of stake tests" =
+
+    (* TODO: expand long fork generation to support relativity constraints
+    let%test_unit "selection case: distinct checkpoints & different min window densities" =
+      failwith "TODO"
+
+    let%test_unit "selection case: distinct checkpoints & equal min window densities & different lengths" =
+      failwith "TODO"
+
+    let%test_unit "selection case: distinct checkpoints & equal min window densities & equal lengths & different vrfs" =
+      failwith "TODO"
+
+    let%test_unit "selection case: distinct checkpoints & equal min window densities & equal lengths & qequals vrfs & different hashes" =
+      failwith "TODO"
+    *)
+
+    let%test_unit "selection invariant: candidate selections are not \
+                   commutative" =
+      let logger = Logger.null () in
+      let constants = Lazy.force Constants.for_unit_tests in
+      let select existing candidate =
+        Hooks.select ~constants ~existing ~candidate ~logger
+      in
+      Quickcheck.test gen_spot_pair
+        ~f:
+          (assert_hashed_consensus_state_pair
+             ~assertion:"chains do not trigger a selection cycle"
+             ~f:(fun (a, b) -> (select a b, select b a) <> (`Take, `Take)))
+
+    (* We define a homogeneous binary relation for consensus states by adapting the binary chain
+     * selection rule and extending it to consider equality of chains. From this, we can test
+     * that this extended relations forms a non-strict partial order over the set of consensus
+     * states.
+     *
+     * We omit partial order reflexivity and antisymmetry tests as they are merely testing properties
+     * of equality related to the partial order we define from binary chain selection. Chain
+     * selection, as is written, will always reject equal elements, so the only property we are
+     * truly interested in it holding is transitivity (when lifted to a homogeneous binary relation).
+     *
+     * TODO: Improved quickcheck generator which better explores related states via our spot
+     * pair generation rules. Doing this requires re-working our spot pair generators to
+     * work by extending an already generated consensus state with some relative constraints.
+     *)
+    let%test_unit "selection invariant: partial order transitivity" =
+      let logger = Logger.null () in
+      let constants = Lazy.force Constants.for_unit_tests in
+      let select existing candidate =
+        Hooks.select ~constants ~existing ~candidate ~logger
+      in
+      let ( <= ) a b =
+        match (select a b, select b a) with
+        | `Keep, `Keep ->
+            true
+        | `Keep, `Take ->
+            true
+        | `Take, `Keep ->
+            false
+        | `Take, `Take ->
+            assert_hashed_consensus_state_pair (a, b)
+              ~assertion:"chains do not trigger a selection cycle"
+              ~f:(Fn.const false) ;
+            (* unreachable *)
+            false
+      in
+      let chains_hold_transitivity (a, b, c) =
```

Comment:
Maybe just as good to change this to

```ocaml
let chains_hold_transitivity (a, b, c) =
  let implies x y = not x || y in
  implies ((a <= b) && (b <= c)) (a <= c)
```
?

I guess your code is basically kind of sorting them to make the implication as non-trivial as possible, but it is hard to check what the code as written is actually doing

Reply:
Yeah, I'm not proud of this function I wrote. In this case, though, if the order was `[a; c; b]` instead of `[a; b; c]`, would this still check that transitivity is held? There is no guarantee on the order of `(a, b, c)` coming into this function, so any permutation of that order is possible.

Reply:
I guess maybe I can sort the list `[a; b; c]` using `(<=)`, and then assert that `not (b <= a) and not (c <= a) and not (c <= b)`? Would that work?